File: data structures and algorithms/python/03_004.py
def merge_sort(array):
  """
  Sorts using merge sort algorithm.
  
  Args:
     array: The list to be sorted
     
  Returns:
     A new sorted list
  """
  # Base case: already sorted
  if len(array) <= 1:
     return array
     
  # Divide array into halves
  mid = len(array) // 2
  
  # Recursively sort halves
  left = merge_sort(array[:mid])
  right = merge_sort(array[mid:])
  
  # Merge sorted halves
  return merge(left, right)

def merge(left, right):
  """
  Merges two sorted arrays.
  
  Args:
     left: First sorted array
     right: Second sorted array
     
  Returns:
     New sorted array with all elements
  """
  result = []
  
  # Compare and merge elements
  while left and right:
     result.append(left.pop(0) if left[0] <= right[0] 
                 else right.pop(0))
  
  # Add remaining elements
  return result + left + right

# Example usage
array = [23, 16, 6, 59, 3, 11, 37]
sorted_array = merge_sort(array)
print(sorted_array)  # [3, 6, 11, 16, 23, 37, 59]

File: data structures and algorithms/cs/03_004.cs
using System;

public class MergeSort
{
   public static int[] MergeSortArray(int[] array)
   {
      if (array.Length <= 1)
         return array;

      int mid = array.Length / 2;
      int[] left = MergeSortArray(
         array[0..mid]);
      int[] right = MergeSortArray(
         array[mid..]);

      return Merge(left, right);
   }

   private static int[] Merge(int[] left, int[] right)
   {
      int[] result = new int[left.Length + right.Length];
      int i = 0, j = 0, k = 0;

      // Merge left and right arrays
      while (i < left.Length && j < right.Length)
         result[k++] = (left[i] <= right[j]) ?
            left[i++] : right[j++];

      // Copy remaining elements of left array
      while (i < left.Length)
         result[k++] = left[i++];

      // Copy remaining elements of right array
      while (j < right.Length)
         result[k++] = right[j++];

      return result;
   }

   public static void Main(string[] args)
   {
      int[] array = { 23, 16, 6, 59, 3, 11, 37 };
      Console.WriteLine(
         string.Join(", ", MergeSortArray(array)));
   }
}